### ISSUE TITLE
FEATURE: allow strings instead of enum for popper placement

### ISSUE THREAD

**Issue by nguyenvuduc** [tonysamperi/ngx-popperjs#49]
**NOTE**
**Describe the bug**
I got this error when using the directive like what the document described:
```
 <popper-content #popper1Content>
     <p class="bold">Popper on bottom</p>
 </popper-content>
 <div [popper]="popper1Content"
      [popperShowOnStart]="true"
      [popperTrigger]="'click'"
      [popperHideOnClickOutside]="true"
      [popperHideOnScroll]="true"
      [popperPlacement]="'bottom'">
     <p class="bold">Hey!</p>
     <p class="thin">Choose where to put your popper!</p>         
 </div>
```

**To Reproduce**
https://stackblitz.com/edit/angular-4i8npp?file=src/app.component.ts

**Expected behavior**
The compilation should be successful.
This failure is because the `ngx-popperjs-placements.model.ts` and `ngx-popperjs-triggers.model.ts` are declared as an enum instead of a string union. I.e.
```
export enum NgxPopperjsTriggers {
  click = "click",
  hover = "hover",
  mousedown = "mousedown",
  none = "none"
}
```

should be

```
export type NgxPopperjsTriggers = 'click'|'hover'|'mousedown'|'none';
```

**Screenshots**
![image](https://user-images.githubusercontent.com/5864656/236702284-e5e54d40-371e-41d7-91a0-43834199aa3f.png)

** Versions
 - ngx-popperjs version [15.1.0]
 - @angular/core version [15.1.2]
 - typescript version [4.7.2]

Thank you for the awesome library.

**Comment by tonysamperi** [comment]
@nguyenvuduc 

1. Use the search
2. No man, it's your code the problem...I can give you a hint if you want...but you should take a look to Angular docs...

**Comment by nguyenvuduc** [comment]
Could you give me a hint?
The best I can find is this issue with strictTemplates of Angular Ivy: https://github.com/angular/angular/issues/35363
- Turning off the strictTemplates is not a good practice for me.
- Declaring a variable of type `NgxPopperjsTriggers`/`NgxPopperjsPlacement` is clumsy
Thanks.

**Comment by tonysamperi** [comment]
@nguyenvuduc sorry man I totally missed your reply.
So you know that strictTemplates is the problem...
Then you'll know that there's no way of bypassing this except maybe creating a wrapper of the directive which accepts the string and then you map the string to the enum...
That way you centralize the fine usage and keep your dirt in the rest of the app...

**Comment by tonysamperi** [comment]
Actually this idea is so good that I think I'll add this version of directive before updating to Angular 16

**Comment by tonysamperi** [comment]
Done in v15.2.0

**Comment by nguyenvuduc** [comment]
Thank @tonysamperi for looking into this. 
I workaround the issue by declaring a service PopperService and injecting it into any component that using Popper.
![image](https://github.com/tonysamperi/ngx-popperjs/assets/5864656/ea646698-4a6c-408d-a91f-a15b7491b5f8)

But your popper-loose directive wrapper is a brilliant idea and much more straightforward. I will update my project.
Thank you so much.

**Comment by k290** [comment]
@tonysamperi  would it be possible to please get a loose version for the "trigger" like you did with the "placement" if you wouldn't mind?

If you don't have time, I'm happy to put up a pull request sometime in a few weeks probably.

**Comment by tonysamperi** [comment]
> @tonysamperi would it be possible to please get a loose version for the "trigger" like you did with the "placement" if you wouldn't mind?
> 
> If you don't have time, I'm happy to put up a pull request sometime in a few weeks probably.

Hi, yes that's an easy one. I'll do in a few days...on both v15 and v16.

Cheers

**Comment by k290** [comment]
> > @tonysamperi would it be possible to please get a loose version for the "trigger" like you did with the "placement" if you wouldn't mind?
> > If you don't have time, I'm happy to put up a pull request sometime in a few weeks probably.
> 
> Hi, yes that's an easy one. I'll do in a few days...on both v15 and v16.
> 
> Cheers

Appreciated, thank you very much.

**Comment by k290** [comment]
I found time to fork the library and have a look at how it works and experimented a bit. If you want type-safety and compile-time errors so that only strings from the enum are allowed with the loose values you can do this:

```  
    @Input()
    set popperLoosePlacement(newValue: `${NgxPopperjsPlacements}`) {
           this.placement = <NgxPopperjsPlacements>newValue;
    }
    
    @Input()
    set popperLooseTrigger(newValue: `${NgxPopperjsTriggers}`) {
       this.showTrigger = <NgxPopperjsTriggers>newValue;
    }
    
```

Additionally, its possible that using this technique could be a non-breaking change to the normal popper that keeps it's type-safety (making the loose one unnecessary). That would be your call. I'm not sure how much testing would be involved. It would look like this:

```
   @Input("popperPlacement")
    set placement(newValue: `${NgxPopperjsPlacements}` | NgxPopperjsPlacements) {
        this._popperPlacement = <NgxPopperjsPlacements>newValue;
        this._checkExisting("placement", <NgxPopperjsPlacements>newValue);
    }
 ```
 
 And a `get` and `set` would have to be made for the `showTrigger` property to do a similar thing, since it does not already have a setter like the `placement` one

**Comment by tonysamperi** [comment]
@k290 about the non breaking for the standard, the fact is that I want the standard to be strict.
I don't want strings to work 😅! Because for MY use , I don't want strings around!
If someone likes having strings like that, I don't see why keeping the strict templates check.
So I created the other one, just because I fair amount (to be honest I'd say less than 10) of issues have been opened to "report" that strings wouldn't work!
But ...after all this is issue #49...if you consider that this lib has 50K downloads a week...basically everyone is fine with how it works!
So I'll keep the loose popper just to make the few happy!

**Comment by tonysamperi** [comment]
Fixed in v16.0.1 + v15.2.1

**Comment by john8329** [comment]
Just a note, it's very easy to just use the enums directly in the template. I don't think it's clear enough so adding this.
```
[popperLoosePlacement]="'bottom'"
[popperPlacement]="NgxPopperjsPlacements.BOTTOM"

[popperLooseTrigger]="'click'"
[popperTrigger]="NgxPopperjsTriggers.click"
```

In the class
```
  protected readonly NgxPopperjsPlacements = NgxPopperjsPlacements;
  protected readonly NgxPopperjsTriggers = NgxPopperjsTriggers;
```
No need for services.

The type system could be programmed to allow strings too but I guess this is fine.
